wai-extra/Network/Wai/Middleware/RequestSizeLimit/Internal.hs

```diff
@@ -0,0 +1,42 @@
+-- | Internal constructors and helper functions. Note that no guarantees are given for stability of these interfaces.
+module Network.Wai.Middleware.RequestSizeLimit.Internal
+    ( RequestSizeLimitSettings(..)
+    ) where
+
+import Network.Wai
+import Data.Word (Word64)
+
+-- | Settings to configure 'requestSizeLimitMiddleware'.
+--
+-- This type and its accessors (but not the constructor) are exported from "Network.Wai.Middleware.RequestSizeLimit".
+-- Since the constructor isn't exported, create a default value with 'defaultRequestSizeLimitSettings' first, then override the individual fields.
+--
+-- If you need to access the constructor directly, it's exported from "Network.Wai.Middleware.RequestSizeLimit.Internal".
+--
+-- ==== __Examples__
+--
+-- ===== Conditionally setting the limit based on the request
+-- > {-# LANGUAGE OverloadedStrings #-}
+-- > import Network.Wai
+-- > import Network.Wai.Middleware.RequestSizeLimit
+-- >
+-- > let megabyte = 1024 * 1024
+-- > let sizeForReq req = if pathInfo req == ["upload", "image"] then pure $ Just $ megabyte * 20 else pure $ Just $ megabyte * 2
+-- > let finalSettings = defaultRequestSizeLimitSettings { maxLengthForRequest = sizeForReq }
+--
+-- ===== JSON response
+-- > {-# LANGUAGE OverloadedStrings #-}
+-- > import Network.Wai
+-- > import Network.Wai.Middleware.RequestSizeLimit
+-- > import Network.HTTP.Types.Status (requestEntityTooLarge413)
+-- > import Data.Aeson
+-- > import Data.Text (Text)
+-- >
+-- > let jsonResponse = \_maxLen _app _req sendResponse -> sendResponse $ responseLBS requestEntityTooLarge413 [("Content-Type", "application/json")] (encode $ object ["error" .= ("request size too large" :: Text)])
+-- > let finalSettings = defaultRequestSizeLimitSettings { onLengthExceeded = jsonResponse }
+--
+-- @since 3.1.1
+data RequestSizeLimitSettings = RequestSizeLimitSettings
+    { maxLengthForRequest :: Request -> IO (Maybe Word64) -- ^ Function to determine the maximum request size in bytes for the request. Return 'Nothing' for no limit. Since 3.1.1
```

Comment:
I realized that before I was only taking a static Word64 parameter, but this would be limiting because it'd be only a single value. Now this is updated to mirror Yesod's interface, which allows constructing the limit dynamically based on the request.

Reply:
This is a good improvement. But I would recommend against exposing these accessors directly. It ends up tying our hands with API changes in the future. I'd recommend only exposing explicit setter functions.

Reply:
Ah ok. I've exposed just explicit setter functions now :+1: